src/System.Memory/src/System/SpanSortHelpers.Common.cs

```diff
@@ -0,0 +1,175 @@
+﻿using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+#if !netstandard
+using Internal.Runtime.CompilerServices;
+#endif
+
+namespace System
+{
+    // TODO: Rename to SpanSortHelpers before move to corefx
+    internal static partial class SpanSortHelpersCommon
+    {
+
+        // This is the threshold where Introspective sort switches to Insertion sort.
+        // Empirically, 16 seems to speed up most cases without slowing down others, at least for integers.
+        // Large value types may benefit from a smaller number.
+        internal const int IntrosortSizeThreshold = 16;
+
+        internal static int FloorLog2PlusOne(int n)
+        {
+            Debug.Assert(n >= 2);
+            int result = 2;
+            n >>= 2;
+            while (n > 0)
+            {
+                ++result;
+                n >>= 1;
+            }
+            return result;
+        }
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Swap<T>(ref T items, int i, int j)
+        {
+            Debug.Assert(i != j);
+            Swap(ref Unsafe.Add(ref items, i), ref Unsafe.Add(ref items, j));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Swap<T>(ref T a, ref T b)
+        {
+            T temp = a;
+            a = b;
+            b = temp;
+        }
+
+        // This started out with just LessThan.
+        // However, due to bogus comparers, comparables etc.
+        // we need to preserve semantics completely to get same result.
+        internal interface IDirectComparer<in T>
+        {
+            bool GreaterThan(T x, T y);
+            bool LessThan(T x, T y);
+            bool LessThanEqual(T x, T y); // TODO: Delete if we are not doing specialize Sort3
+        }
+        //
+        // Type specific DirectComparer(s) to ensure optimal code-gen
+        //
+        internal struct SByteDirectComparer : IDirectComparer<sbyte>
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Are all these AggressiveInlining attributes necessary? I think we should remove them and only use them sparingly where it really matters and we know that the JIT can't do it. At least, that is the guideline I have been given.

cc @jkotas, @AndyAyersMS

Reply:
@ahsonkhan I'll await feedback before deleting them, but deleting `LessThanEqual` since not needed.

Reply:
My guess would be that most of these attributes are not needed -- if a method simply invokes another method it is quite likely the jit will inline it without the attribute.

Overall guidance is that you should consider adding the `AggressiveInlining` attribute only if both of the following are true:
* use of the attribute leads to large performance improvements in important scenarios
* use of the attribute does not create large or unbounded amounts of code growth

In particular you need to be careful when using it in combination with generics as the runtime can create large numbers of generic instantiations and the benefits of inlining are often type specific.